src/Servers/IIS/IIS/src/Core/IISHttpContext.cs

```diff
@@ -615,10 +615,24 @@ protected void ReportApplicationError(Exception ex)
         Log.ApplicationError(_logger, ((IHttpConnectionFeature)this).ConnectionId, ((IHttpRequestIdentifierFeature)this).TraceIdentifier, ex);
     }
 
+    // This method should only be called from the end of request processing, this lets use avoid most
+    // of the rentrancy issues with using indicate completion.
     public void PostCompletion(NativeMethods.REQUEST_NOTIFICATION_STATUS requestNotificationStatus)
     {
-        NativeMethods.HttpSetCompletionStatus(_requestNativeHandle, requestNotificationStatus);
-        NativeMethods.HttpPostCompletion(_requestNativeHandle, 0);
+        // We post completion to the thread pool instead of the IIS thread pool to avoid having
+        // new requests compete with existing threads for IIS threads.
+        // We're going to use the local queue so that we can resume the state machine on this thread after
+        // unwinding.
+        ThreadPool.UnsafeQueueUserWorkItem(static state =>
+        {
+            var (handle, status) = state;
```

Comment:
nit: There doesn't seem to be any benefit in making the callback static and passing in state vs a closure. The tuple will allocate when it's boxed to `object state`.

Not saying it should change, just curious whether there is a benefit I don't see, or habit of passing in state

Reply:
> nit: There doesn't seem to be any benefit in making the callback static and passing in state vs a closure. The tuple will allocate when it's boxed to object state.

This uses the generic overload https://docs.microsoft.com/en-us/dotnet/api/system.threading.threadpool.unsafequeueuserworkitem?view=net-6.0#system-threading-threadpool-unsafequeueuserworkitem-1(system-action((-0))-0-system-boolean). No boxing 😄